src/main/scala/org/ergoplatform/nodeView/wallet/scanning/ScanningPredicate.scala

```diff
@@ -0,0 +1,50 @@
+package org.ergoplatform.nodeView.wallet.scanning
+
+import org.ergoplatform.ErgoBox
+import sigmastate.Values
+
+/**
+  * Basic interface for box scanning predicate functionality
+  *
+  * See EIP-0001 for details (https://github.com/ergoplatform/eips/blob/master/eip-0001.md)
+  */
+sealed trait ScanningPredicate {
+  def filter(box: ErgoBox): Boolean
+}
+
+case class ContainsScanningPredicate(regId: ErgoBox.RegisterId, bytes: Array[Byte]) extends ScanningPredicate {
+
+  override def filter(box: ErgoBox): Boolean = {
+    box.get(regId).exists {
+      _ match {
+        case Values.ByteArrayConstant(arr) => arr.toArray.containsSlice(bytes)
+        case _ => false
+      }
+    }
+  }
+}
+
+case class EqualsScanningPredicate(regId: ErgoBox.RegisterId, bytes: Array[Byte]) extends ScanningPredicate {
+  override def filter(box: ErgoBox): Boolean = {
+    box.get(regId).exists {
+      _ match {
+        case Values.ByteArrayConstant(arr) => arr.toArray.sameElements(bytes)
+        case _ => false
+      }
+    }
+  }
+}
+
+case class ContainsAssetPredicate(assetId: ErgoBox.TokenId) extends ScanningPredicate {
+  override def filter(box: ErgoBox): Boolean = {
+    box.additionalTokens.exists(_._1.sameElements(assetId))
+  }
+}
+
+case class AndScanningPredicate(subPredicates: ScanningPredicate*) extends ScanningPredicate {
+  override def filter(box: ErgoBox): Boolean = subPredicates.forall(p => p.filter(box))
```

Comment:
This is hotspot code.
Please see [SPSG](https://github.com/ScorexFoundation/sigmastate-interpreter/blob/master/docs/perf-style-guide.md#looping-using-for) as motivation for replacing 
`forall` with explicit loop using `cfor`.